### PR TITLE
OpenSSL SHA hashing implementation.

### PR DESCRIPTION
**Pull request by pkthapa**
This PR includes below changes:-
* OpenSSL library implementation for `SHA_256` & `SHA3_256` hashing.
* Design change for existing Crypto++ library hashing (existing `Digest` class modified). The storing of generated digest and the digest generation procedures are decoupled and achieved with the help of `DigestHolder` & `DigestCreator` classes respectively.
* Introduced compilation macros to choose between cryptographic hashing libraries i.e. Cryptopp and OpenSSL (SHA_256 and SHA3_256).

Note: Previous PR [closed]: https://github.com/arc-vmware/concord-bft/pull/1

This PR still has Crypto++ library for digest generation.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -113,6 +113,25 @@ if(CODECOVERAGE)
     message( "-- Building with llvm Code Coverage Tools")
 endif()
 
+if((NOT USE_CRYPTOPP_HASH) AND (NOT USE_OPENSSL_SHA_256) AND (NOT USE_OPENSSL_SHA3_256))
+    message(FATAL_ERROR "None of the cryptographic hashing libraries are enabled.")
+elseif(USE_OPENSSL_SHA_256 AND USE_OPENSSL_SHA3_256)
```
Comment (arc-vmware) [inline CMakeLists.txt]: Why is this check needed? As per the code, it seems that we are using a prioritized approach.
#if defined USE_CRYPTOPP_HASH
using Digest = DigestHolder<CryptoppDigestCreator>;
#elif defined USE_OPENSSL_SHA_256
using Digest = DigestHolder<OpenSSLDigestCreator<SHA2_256> >;
#elif defined USE_OPENSSL_SHA3_256
using Digest = DigestHolder<OpenSSLDigestCreator<SHA3_256> >;
#endif
So if everything is enabled then we set Cryptopp. So I think setting all the macros ON is not an issue as such. But keeping this logic will add more maintenance overhead in the future. So we will just set one single in our build env, and we are tacking this as per priority. Only the first check is good enough as we need at least one to be kept.

File: CMakeLists.txt [hunk 2]
```diff
@@ -113,6 +113,25 @@ if(CODECOVERAGE)
...
+    message(FATAL_ERROR "Cannot have OpenSSL's both SHA_256 and SHA3_256 hashing enabled.")
+elseif((USE_OPENSSL_SHA_256 AND USE_CRYPTOPP_HASH) OR (USE_OPENSSL_SHA3_256 AND USE_CRYPTOPP_HASH))
+    message(FATAL_ERROR "Cannot have both Cryptopp and OpenSSL hashing enabled.")
+endif()
+
+if(USE_OPENSSL_SHA_256)
```
Comment (arc-vmware) [inline CMakeLists.txt]: This set of if will make sure that we set only one, so the above checks are redundant.

File: CMakeLists.txt [hunk 3]
```diff
@@ -113,6 +113,25 @@ if(CODECOVERAGE)
...
+    message("-- USE_OPENSSL_SHA_256 Enabled")
+    string(APPEND CMAKE_CXX_FLAGS " -DUSE_OPENSSL_SHA_256")
+elseif(USE_OPENSSL_SHA3_256)
+    message("-- USE_OPENSSL_SHA3_256 Enabled")
+    string(APPEND CMAKE_CXX_FLAGS " -DUSE_OPENSSL_SHA3_256")
+elseif(USE_CRYPTOPP_HASH)
+    message("-- USE_CRYPTOPP_HASH Enabled")
+    string(APPEND CMAKE_CXX_FLAGS " -DUSE_CRYPTOPP_HASH")
+endif()
```
Comment (arc-vmware) [inline CMakeLists.txt:127]: Add a else for the case if none of these are defined and do FATAL_ERROR and remove the if((NOT USE_CRYPTOPP_HASH) AND (NOT USE_OPENSSL_SHA_256) AND (NOT USE_OPENSSL_SHA3_256)) in line no. 116

File: Makefile [hunk 4]
```diff
@@ -88,8 +92,10 @@ CONCORD_BFT_CMAKE_FLAGS?= \
 			-DLEAKCHECK=${CONCORD_BFT_CMAKE_ASAN} \
 			-DTHREADCHECK=${CONCORD_BFT_CMAKE_TSAN} \
 			-DCODECOVERAGE=${CONCORD_BFT_CMAKE_CODECOVERAGE} \
-			-DTXN_SIGNING_ENABLED=${CONCORD_BFT_CMAKE_TRANSACTION_SIGNING_ENABLED}
-
+			-DTXN_SIGNING_ENABLED=${CONCORD_BFT_CMAKE_TRANSACTION_SIGNING_ENABLED} \
+			-DUSE_OPENSSL_SHA_256=${CONCORD_BFT_CMAKE_USE_OPENSSL_SHA_256} \
```
Comment (arc-vmware) [inline Makefile:96]: as we see our #defines is:
#if defined USE_CRYPTOPP_HASH
using Digest = DigestHolder<CryptoppDigestCreator>;
#elif defined USE_OPENSSL_SHA_256
using Digest = DigestHolder<OpenSSLDigestCreator<SHA2_256> >;
#elif defined USE_OPENSSL_SHA3_256
using Digest = DigestHolder<OpenSSLDigestCreator<SHA3_256> >;
#endif

So if a macro is set to FALSE, is it defined or not?

File: util/include/DigestImpl.ipp [hunk 5]
```diff
@@ -133,8 +139,6 @@ class OpenSSLDigestCreator : public DigestCreator {
   void final() override {}
 
  private:
-  SHACTX hash_ctx_;
+  EVP_MD_CTX* hash_ctx_;
```
Comment (arc-vmware) [inline util/include/DigestImpl.ipp]: This should be SHACTX and not EVP_MD_CTX*.

File: util/include/DigestImpl.ipp [hunk 6]
```diff
@@ -133,8 +139,6 @@ class OpenSSLDigestCreator : public DigestCreator {
...
 };
-}  // namespace concord::util::digest
-
-//#include "DigestImpl.ipp"
\ No newline at end of file
+}  // namespace concord::util::digest
```
Comment (arc-vmware) [inline util/include/DigestImpl.ipp]: Better to add a new line.

File: util/src/DigestImpl.cpp [hunk 7]
```diff
@@ -89,4 +89,4 @@ DigestUtil::Context::~Context() {
     internalState = NULL;
   }
 }
-}  // namespace concord::util::digest
+}  // namespace concord::util::digest
```
Comment (arc-vmware) [inline util/src/DigestImpl.cpp]: Add a new line.

**Review by arc-vmware — CHANGES_REQUESTED** [review]
Good code, left some comments.

File: Makefile (hunk 4 above)
Comment (pkthapa) [reply Makefile:96]: If all the three macros are set to false, then we get compilation stops during `make format` check.

File: CMakeLists.txt (hunk 2 above)
Comment (pkthapa) [reply CMakeLists.txt]: Addressed https://github.com/arc-vmware/concord-bft/pull/2/commits/be7b4d3f850359b235887ae82cbe8f4c29a080c4

File: CMakeLists.txt (hunk 3 above)
Comment (pkthapa) [reply CMakeLists.txt:127]: Addressed https://github.com/arc-vmware/concord-bft/pull/2/commits/be7b4d3f850359b235887ae82cbe8f4c29a080c4

File: CMakeLists.txt (hunk 1 above)
Comment (pkthapa) [reply CMakeLists.txt]: Addressed https://github.com/arc-vmware/concord-bft/pull/2/commits/be7b4d3f850359b235887ae82cbe8f4c29a080c4

File: Makefile (hunk 4 above)
Comment (pkthapa) [reply Makefile:96]: Addressed https://github.com/arc-vmware/concord-bft/pull/2/commits/be7b4d3f850359b235887ae82cbe8f4c29a080c4

File: util/include/DigestImpl.ipp (hunk 6 above)
Comment (pkthapa) [reply util/include/DigestImpl.ipp]: Addressed https://github.com/arc-vmware/concord-bft/pull/2/commits/be7b4d3f850359b235887ae82cbe8f4c29a080c4

File: util/src/DigestImpl.cpp (hunk 7 above)
Comment (pkthapa) [reply util/src/DigestImpl.cpp]: Addressed https://github.com/arc-vmware/concord-bft/pull/2/commits/be7b4d3f850359b235887ae82cbe8f4c29a080c4

File: util/include/Digest.hpp [hunk 8]
```diff
@@ -22,13 +22,16 @@ namespace concord::util::digest {
...
-#if defined USE_CRYPTOPP
+#if defined USE_CRYPTOPP_HASH
 using Digest = DigestHolder<CryptoppDigestCreator>;
-#elif defined USE_OPENSSL_SHA256
-using Digest = DigestHolder<OpenSSLDigestCreator<EVP_sha256> >;
-#elif defined USE_OPENSSL_SHA3_256
-using Digest = DigestHolder<OpenSSLDigestCreator<EVP_sha3_256> >;
+#endif
+
+#if defined USE_OPENSSL_SHA_256
```
Comment (arc-vmware) [inline util/include/Digest.hpp]: why is elif removed?

File: util/include/DigestImpl.ipp (hunk 5 above)
Comment (pkthapa) [reply util/include/DigestImpl.ipp]: Changed to SHACTX.

File: util/include/Digest.hpp (hunk 8 above)
Comment (pkthapa) [reply util/include/Digest.hpp]: Code reverted.

**Review by arc-vmware — APPROVED** [review]
LGTM !!!

**Pull request merged**